lib/Transforms/Scalar/Scalarizer.cpp

```diff
@@ -673,13 +675,47 @@ bool Scalarizer::finish() {
   if (Gathered.empty())
     return false;
   // HLSL Change Begins.
+  Module &M = *Gathered.front().first->getModule();
+  LLVMContext &Ctx = M.getContext();
+  const DataLayout &DL = M.getDataLayout();
+  bool HasDbgInfo = getDebugMetadataVersionFromModule(M) != 0;
   // Map from an extract element inst to a Value which replaced it.
   DenseMap<Instruction *, Value*> EltMap;
   // HLSL Change Ends.
   for (GatherList::iterator GMI = Gathered.begin(), GME = Gathered.end();
        GMI != GME; ++GMI) {
     Instruction *Op = GMI->first;
     ValueVector &CV = *GMI->second;
+    // HLSL Change Begin - debug info in scalarizer.
+    if (HasDbgInfo) {
+      if (auto *L = LocalAsMetadata::getIfExists(Op)) {
+        if (auto *DINode = MetadataAsValue::getIfExists(Ctx, L)) {
+          Type *Ty = Op->getType();
+          unsigned Count = Ty->getVectorNumElements();
+          Type *EltTy = Ty->getVectorElementType();
+          unsigned EltSize = DL.getTypeSizeInBits(EltTy);
+          for (User *U : DINode->users())
+            if (DbgValueInst *DVI = dyn_cast<DbgValueInst>(U)) {
+              DIBuilder DIB(M, /*AllowUnresolved*/ false);
+              auto *VarInfo = DVI->getVariable();
+              DebugLoc DbgLoc = DVI->getDebugLoc();
+              unsigned Offset = 0;
+              for (unsigned I = 0; I < Count; ++I) {
+                // TODO: need to use DIExpression::createFragmentExpression for
```

Comment:
Can you put an assert instead?

Reply:
It will hit the assert a lot.
DIExpression::createFragmentExpression is not exist in llvm 3.7, I will fix it in another pull request.